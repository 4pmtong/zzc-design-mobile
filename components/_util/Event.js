const EVENT_NAME_MAP = {
    transitionend: {
        transition: 'transitionend',
        WebkitTransition: 'webkitTransitionEnd',
        MozTransition: 'mozTransitionEnd',
        OTransition: 'oTransitionEnd',
        msTransition: 'MSTransitionEnd'
    },

    animationend: {
        animation: 'animationend',
        WebkitAnimation: 'webkitAnimationEnd',
        MozAnimation: 'mozAnimationEnd',
        OAnimation: 'oAnimationEnd',
        msAnimation: 'MSAnimationEnd'
    }
};

const endEventsTransitionend = [];
const endEventsAnimationend = [];

function detectEvents() {
    // 判断是否支持AnimationEvent
    if ( 'AnimationEvent' in window ) {
        delete EVENT_NAME_MAP.animationend.WebkitAnimation;
    }
    // 判断是否支持TransitionEvent
    if ( 'TransitionEvent' in window ) {
        delete EVENT_NAME_MAP.transitionend.WebkitTransition;
    }

    Object.keys( EVENT_NAME_MAP.animationend ).forEach( function ( key ) {
        endEventsAnimationend.push( EVENT_NAME_MAP.animationend[key] );
    } );

    Object.keys( EVENT_NAME_MAP.transitionend ).forEach( function ( key ) {
        endEventsTransitionend.push( EVENT_NAME_MAP.transitionend[key] );
    } );
}

function addEventListener( node, eventName, eventListener, _this ) {
    if ( _this ) {
        node.addEventListener( eventName, eventListener.bind( _this ), false );
    } else {
        node.addEventListener( eventName, eventListener, false );
    }
}

function removeEventListener( node, eventName, eventListener, _this ) {
    if ( _this ) {
        node.removeEventListener( eventName, eventListener.bind( _this ), false );
    } else {
        node.removeEventListener( eventName, eventListener, false );
    }
}

const transitionEvents = {
    addEndEventListener( node, eventListener, _this ) {
        if ( endEventsTransitionend.length === 0 ) {
            window.setTimeout( eventListener, 0 );
            return;
        }

        _this = _this ? _this : null;

        endEventsTransitionend.forEach( ( endEvents ) => {
            addEventListener( node, endEvents, eventListener, _this );
        } );
    },
    removeEndEventListener( node, eventListener, _this ) {
        if ( endEventsTransitionend.length === 0 ) {
            return;
        }
        endEventsTransitionend.forEach( ( endEvent ) => {
            removeEventListener( node, endEvent, eventListener, _this );
        } );
    }
};

const animationEvents = {

    addEndEventListener( node, eventListener, _this ) {
        if ( endEventsAnimationend.length === 0 ) {
            window.setTimeout( eventListener, 0 );
            return;
        }
        _this = _this ? _this : null;

        endEventsAnimationend.forEach( ( endEvents ) => {
            addEventListener( node, endEvents, eventListener, _this );
        } );
    },
    removeEndEventListener( node, eventListener, _this ) {
        if ( endEventsAnimationend.length === 0 ) {
            return;
        }
        endEventsAnimationend.forEach( ( endEvent ) => {
            removeEventListener( node, endEvent, eventListener, _this );
        } );
    }
};

if ( typeof window !== 'undefined' && typeof document !== 'undefined' ) {
    detectEvents();
}

export {
    transitionEvents,
    animationEvents
};