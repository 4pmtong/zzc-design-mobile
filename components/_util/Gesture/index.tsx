import React from 'react';

export declare type Finger = {
    x: number // e.touches[i].pageX
    y: number // e.touches[i].pageY
};

export interface GestureProps {
    children: JSX.Element
}

export interface IGestureStatus {
    touches?: Finger[],
    startTouches?: Finger[]
    preTouches?: Finger[] | null
}

export default class Gesture extends React.PureComponent<GestureProps> {
    protected gesture: IGestureStatus

    now (): number {
        return new Date().getTime();
    }

    initGestureStatus ( e: any ): void {
        const touches = this.getTouches( e );
        const now = this.now();
        this.setGestureStatus( {
            touches,
            startTime: now,
            time: now,
            startTouches: touches,
            preTouches: null
        } );
    }

    clearGestureStatus (): void {
        delete this.gesture;
    }

    updateGestureStatus ( e: any ): void {
        this.gesture = {
            touches: this.getTouches( e ),
            preTouches: this.gesture.touches
        };
    }

    setGestureStatus ( params: any ): void {
        if ( !this.gesture ) {
            this.gesture = {};
        }
        // 记录上一个touches信息
        if ( this.gesture.touches ) {
            this.gesture.preTouches = this.gesture.touches;
        }
        this.gesture = {
            ...this.gesture,
            ...params
        };
    }

    getTouches ( e: any ): any {
        return Array.prototype.slice.call( e.touches ).map( item => ( {
            x: item.screenX,
            y: item.screenY
        } ) );
    }

    _handleTouchStart ( e ) {
        this.clearGestureStatus();
        if ( e.touches.length > 1 ) {
            e.preventDefault();
        }
        this.initGestureStatus( e );
    }
    _handleTouchMove ( e ) {
        alert( e.touches.length );
        if ( e.touches.length > 1 ) {
            e.preventDefault();
        }
        this.updateGestureStatus( e );
        console.log( this.gesture );
    }
    _handleTouchCancel ( e ) {
        if ( e.touches.length > 1 ) {
            e.preventDefault();
        }
        this.updateGestureStatus( e );
    }
    _handleTouchEnd ( e ) {
        if ( e.touches.length > 1 ) {
            e.preventDefault();
        }
        this.updateGestureStatus( e );
    }

    render () {
        const { children } = this.props;
        const child = React.Children.only( children );
        const events = {
            onTouchStart: ( e ) => { this._handleTouchStart( e ); },
            onTouchMove: ( e ) => { this._handleTouchMove( e ); },
            onTouchCancel: ( e ) => { this._handleTouchCancel( e ); },
            onTouchEnd: ( e ) => { this._handleTouchEnd( e ); }
        };
        return React.cloneElement( child, {
            ...events
        } );
    }
}