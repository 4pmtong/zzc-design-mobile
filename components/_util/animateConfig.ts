export default {
    'zzc-slide-bottom': {
        enter: 'zzc-slide-bottom-enter',
        enterActive: 'zzc-slide-bottom-enter-active',
        leave: 'zzc-slide-bottom-leave',
        leaveActive: 'zzc-slide-bottom-leave-active'
    },
    'zzc-slide-top': {
        enter: 'zzc-slide-top-enter',
        enterActive: 'zzc-slide-top-enter-active',
        leave: 'zzc-slide-top-leave',
        leaveActive: 'zzc-slide-top-leave-active'
    },
    'zzc-slide-left': {
        enter: 'zzc-slide-left-enter',
        enterActive: 'zzc-slide-left-enter-active',
        leave: 'zzc-slide-left-leave',
        leaveActive: 'zzc-slide-left-leave-active'
    },
    'zzc-slide-right': {
        enter: 'zzc-slide-right-enter',
        enterActive: 'zzc-slide-right-enter-active',
        leave: 'zzc-slide-right-leave',
        leaveActive: 'zzc-slide-right-leave-active'
    },
    'zzc-zoom': {
        enter: 'zzc-zoom-enter',
        enterActive: 'zzc-zoom-enter-active',
        leave: 'zzc-zoom-leave',
        leaveActive: 'zzc-zoom-leave-active'
    },
    'zzc-fade': {
        enter: 'zzc-fade-enter',
        enterActive: 'zzc-fade-enter-active',
        leave: 'zzc-fade-leave',
        leaveActive: 'zzc-fade-leave-active'
    }
};