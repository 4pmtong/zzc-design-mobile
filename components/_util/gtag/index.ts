
import config from './config';

let gtagStatus = 'no-install';
let isInit = false;
const useList = [];
export function initGtag( componentName ) {
    window['dataLayer'] = window['dataLayer'] || [];
    window['gtag'] = function() {
        window['dataLayer'].push( arguments );
    };
    if ( true ) {
    // if ( config.isOpen ) {

        if ( !isInit ) {
            const scriptList = document.getElementsByTagName( 'script' );
            for ( let i = 0; i < scriptList.length; i++ ) {
                if ( scriptList[i].src.indexOf( 'www.googletagmanager.com/gtag/js' ) != -1 ) {
                    gtagStatus = 'install';
                    break;
                }
            }
            if ( gtagStatus == 'no-install' ) {
                const script = document.createElement( 'script' );
                script.src = config.url;
                script.async = true;
                document.body.appendChild( script );
            }
            window['gtag']( 'js', new Date() );
            window['gtag']( 'config', config.id, {
                page_title: document.title,
                page_location: window.location.host,
                page_path: window.location.pathname,
                custom_map: {
                    dimension1: 'componentName',
                    dimension2: 'componentVersion',
                    dimension3: 'componentUseLocation',
                    dimension4: 'componentUseHost',
                    dimension5: 'errorComponentName',
                    dimension6: 'errorInfo',
                    dimension7: 'errorPath',
                    metric1: 'count',
                    metric2: 'errorCount'
                }
            } );
            window['gtag'](
                'event',
                'screen_view',
                {
                    send_to: config.id,
                    screen_name: window.location.host + window.location.pathname,
                    app_name: config.name,
                    app_id: config.name,
                    app_version: config.version
                }
            );
            isInit = true;
        }
    }
}

export function zzcComponentUse ( componentName, action ) {
    if ( useList.indexOf( componentName ) == -1 ) {
        useList.push( componentName );
        window['gtag'](
            'event',
            action,
            {
                send_to: config.id,
                componentName,
                componentVersion: config.version,
                componentUseLocation: `${window.location.host}${window.location.pathname}`,
                componentUseHost: window.location.host,
                count: 1
            }
        );

        window['gtag'](
            'event',
            action,
            {
                send_to: config.id,
                event_category: `${componentName}-${config.version}`,
                event_label: config.version
            }
        );
    }
}
export function zzcComponetError ( componentName, error, info ) {
    window['gtag'](
        'event',
        'component_error',
        {
            send_to: config.id,
            errorComponentName: componentName,
            errorInfo: `${error}-${info.componentStack}`,
            errorPath: window.location.href,
            errorCount: 1
        }
    );
}

