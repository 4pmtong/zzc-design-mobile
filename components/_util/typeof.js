function isType( s, typeString ) {
    return Object.toString.call( s ) === `[object ${typeString}]`;
}

export function isWindow( s ) {
    return isType( s, 'Window' );
}

export function isObject( s ) {
    return isType( s, 'Object' );
}
export function isArray( s ) {
    return Array.isArray ? Array.isArray( s ) : isType( s, 'Array' );
}
export function isString( s ) {
    return isType( s, 'String' );
}
export function isFunction( s ) {
    return isType( s, 'Function' );
}
export function isNumber( s ) {
    return isType( s, 'Number' );
}
export function isRegExp( s ) {
    return isType( s, 'RegExp' );
}
export function isUndefined( s ) {
    return isType( s, 'Undefined' );
}
