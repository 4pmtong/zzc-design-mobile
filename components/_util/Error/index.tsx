import React, { PureComponent } from 'react';
import { zzcComponetError } from '../gtag';

export default function errorComponent( WrappedComponent, componentName ) {
    return class HLError extends PureComponent {
        componentDidCatch( error, info ) {
            zzcComponetError( componentName, error, info );
        }
        render() {
            return (
                <WrappedComponent {...this.props} />
            );
        }
    };
}