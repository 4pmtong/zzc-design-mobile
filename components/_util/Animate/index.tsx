
import React from 'react';
import ReactDOM from 'react-dom';
import { isObject, isDOM } from '../typeof';
import { animationEvents } from '../Event';
import { addClass, removeClass, hasClass } from '../class';

export interface AnimateProps {
    animationName?: string | object,
    visible: boolean,
    onEnd?: Function
}

export default class Animate extends React.PureComponent<AnimateProps> {
    static defaultProps = {
        animationName: null,
        visible: false,
        onEnd: () => { }
    }

    componentDidMount (): void {
        this.addAnimationEvent();
        this.runHookEvent( true );
    }

    componentDidUpdate (): void {
        this.runHookEvent( false );
    }

    // 执行触发钩子事件
    runHookEvent ( isFirst: boolean ): void {
        if ( this.props.visible ) {
            this.enterEvent();
        } else if ( !isFirst ) {
            this.leaveEvent();
        }
    }

    // 绑定动画事件
    addAnimationEvent (): void {
        const node = ReactDOM.findDOMNode( this );
        if ( isDOM( node ) ) {
            animationEvents.addEndEventListener( node, this.animationEvent, this );
        }
    }

    // 进入动画钩子
    enterEvent (): void {
        const { animationName }: any = this.props;
        const node = ReactDOM.findDOMNode( this );
        if ( isDOM( node ) ) {
            if ( isObject( animationName ) ) {
                const { enter, enterActive } = animationName;
                addClass( node, enter );
                addClass( node, enterActive );
            } else {
                addClass( node, animationName );
            }
        }
    }

    // 离开动画钩子
    leaveEvent (): void {
        const { animationName }: any = this.props;
        const node = ReactDOM.findDOMNode( this );
        if ( isDOM( node ) ) {
            if ( isObject( animationName ) ) {
                const { leave, leaveActive } = animationName;
                addClass( node, leave );
                addClass( node, leaveActive );
            } else {
                removeClass( node, animationName );
            }
        }
    }

    // 动画完成触发事件
    animationEvent ( event ): void {
        event.stopPropagation();
        const { animationName }: any = this.props;
        const { enter, enterActive } = animationName;
        const node = ReactDOM.findDOMNode( this );
        if ( isDOM( node ) ) {
            let animationType = 'enter';
            // 当传入object会清楚enter钩子，如果只是传入一个字符串，会保留class直至关闭才去掉class
            if ( isObject( animationName ) ) {
                if ( hasClass( node, enter ) ) {
                    removeClass( node, enter );
                    removeClass( node, enterActive );
                } else {
                    animationType = 'leave';
                }
            } else if ( hasClass( node, animationName ) && !!this.props.visible ) {
                animationType = 'leave';
            }
            this.props.onEnd && this.props.onEnd( animationType );
        }
    }

    render (): any {
        const { children } = this.props;
        return ( children );
    }
}