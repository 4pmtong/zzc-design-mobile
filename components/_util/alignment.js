

import { isUndefined } from './typeof';

export default class Alignment {
    constructor () {
        this.list = [];
    }
    save ( data ) {
        this.list.push( data );
    }
    check () {
        return this.list;
    }
    get ( index ) {
        if ( index === 'first' ) {
            return this.list.shift();
        } else if ( index === 'last' ) {
            return this.list.pop();
        } else if ( !isUndefined( this.list[index] ) ) {
            const result = this.list[index];
            this.list.splice( index, 1 );
            return result;
        }
        return undefined;
    }
}