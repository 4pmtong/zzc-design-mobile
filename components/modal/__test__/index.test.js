import React from 'react';
import { render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import { isFunction } from '../../_util/typeof';

import Modal from '../index.jsx';


test( 'Modal', () => {
  const wrapper = render( <div>
    <Modal
      title="标题"
      closable={true}
      visible={true}
      closeCallback={() => { console.log( 123 ); }}
    >
      <div className="modal-text">这是一个文案...</div>
    </Modal>
  </div> );
  expect( renderToJson( wrapper ) ).toMatchSnapshot();
} );

