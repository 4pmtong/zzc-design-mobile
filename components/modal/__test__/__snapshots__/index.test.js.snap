// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal 1`] = `
<div>
  <div
    class="zzc-dialog"
  >
    <div
      class="zzc-dialog-mask"
    />
    <div
      class="zzc-dialog-box zzc-modal-box"
    >
      <div
        class="zzc-modal-header"
      >
        标题
      </div>
      <div
        class="zzc-modal-body"
      >
        <div
          class="modal-text"
        >
          这是一个文案...
        </div>
      </div>
      <div
        class="zzc-modal-footer"
      >
        <a
          class="zzc-button zzc-button-full zzc-button-default zzc-button-noborder"
        >
          关闭
        </a>
      </div>
    </div>
  </div>
</div>
`;
